addons/account/tests/test_payment.py

```diff
@@ -592,3 +595,99 @@ def test_partial_payment_inv_foreign_payment_domestic(self):
+            'payment_type': 'inbound',
+            'partner_type': 'customer',
+            'amount': 1000,
+            'journal_id': bank_journal_a.id,
+            'partner_id': partner.id,
+        })
+        company_b = self.env['res.company'].create({'name': 'Company B'})
+        bank_journal_b = self.env['account.journal'].create({
+            'name': 'Journal Company B',
+            'type': 'bank',
+            'code': 'b',
+            'company_id': company_b.id
+        })
+        bank_account_b = self.env["res.partner.bank"].create({
+            "acc_number": "	BE68539007547034",
+            "bank_id": self.bank.id,
+            "partner_id": partner.id,
+            "company_id": company_b.id,
+        })
+        payment_b = self.env['account.payment'].create({
+            'payment_date': time.strftime('%Y') + '-01-01',
+            'payment_method_id': self.payment_method_manual_in.id,
+            'payment_type': 'inbound',
+            'partner_type': 'customer',
+            'amount': 1000,
+            'journal_id': bank_journal_b.id,
+            'partner_id': partner.id,
+        })
+
+        # Case 1 : partner has a commercial partner with a bank account from company B
+        bank_account_b.write({
+            'partner_id': partner_b.id
+        })
+        partner.write({
+            'commercial_partner_id': partner_b.id
+        })
+        trigger_onchange_partner(payment_a)
+        self.assertFalse(payment_a.partner_bank_account_id.id)
+        trigger_onchange_partner(payment_b)
+        self.assertEqual(payment_b.partner_bank_account_id.id, bank_account_b.id)
+
+        # Case 2 : partner has a bank account from company B
+        bank_account_b.write({
+            'partner_id': partner.id
+        })
+        trigger_onchange_partner(payment_a)
+        self.assertFalse(payment_a.partner_bank_account_id.id)
+        trigger_onchange_partner(payment_b)
+        self.assertEqual(payment_b.partner_bank_account_id.id, bank_account_b.id)
+
+        # Case 3 : the payment has an invoice from company B with a bank account
+        payment_a.write({
+            'invoice_ids': [
+                (0, 0, {
+                    'company_id': company_b,
+                    'invoice_partner_bank_id': bank_account_b,
+                    'type': 'out_invoice',
+                    'partner_id': partner,
+                    'invoice_date': time.strftime('%Y') + '-01-01',
+                    'date': time.strftime('%Y') + '-01-01',
+                    'invoice_line_ids': [
+                        (0, 0, {'product_id': self.product.id, 'quantity': 1, 'price_unit': 10})
+                    ],
+                })
+            ],
+        })
+        trigger_onchange_partner(payment_a)
+        self.assertFalse(payment_a.partner_bank_account_id.id)
+        trigger_onchange_partner(payment_b)
+        self.assertEqual(payment_b.partner_bank_account_id.id, bank_account_b.id)
```

Comment:
So you always expect the same results?
What about putting that into a function instead of having that helper to trigger the onchange?